spanner/spannertest/db.go

```diff
@@ -617,6 +814,79 @@ func (d *database) ReadAll(tbl spansql.ID, cols []spansql.ID, limit int64) (*raw
 	})
 }
 
+func (d *database) IndexRead(tbl, idx spansql.ID, cols []spansql.ID, keys []*structpb.ListValue, keyRanges keyRangeList, limit int64) (rowIter, error) {
+	return d.readIndex(tbl, idx, cols, func(i *index, ri *rawIter, colIndexes []int) error {
+		// "If the same key is specified multiple times in the set (for
+		// example if two ranges, two keys, or a key and a range
+		// overlap), Cloud Spanner behaves as if the key were only
+		// specified once."
+		done := make(map[*indexRow]bool) // row numbers we've included in ri.
+
+		add := func(row *indexRow) bool {
+			if row == nil || done[row] {
+				return true
+			}
+			done[row] = true
+			for _, r := range row.rows {
+				if limit > 0 && len(ri.rows) >= int(limit) {
+					return false
+				}
+				ri.add(r, colIndexes)
+			}
+			return true
+		}
+		// Specific keys.
+		for _, key := range keys {
+			indexKey, err := i.key(key.Values)
+			if err != nil {
+				return err
+			}
+			// Not an error if the key does not exist.
+			row := i.rowForKey(indexKey)
+			if !add(row) {
+				return nil
```

Comment:
This implementation means that specific keys will take precedence over key ranges. That is not consistent with the specification of the Read method of Cloud Spanner. The simplest example of when this would go wrong is if you execute a read operation with a limit of 1 and you specify both a single key and a key range. The single key is the row that always will be returned (assuming it exists), even if the key range starts at a lower value than the single key. This in other words does not respect the ordering of the entire result set when both single keys and key ranges have been specified.